3단계 - 로또(자동) PR 요청 드립니다.

yoonmin-kim opened the pull request:
안녕하세요 리뷰어님^^
주말 잘 보내고 계시나요?

이번 로또(자동) 기능은 다음과 같은 부분에 신경을 써서 구현하였습니다.

- 배열보다는 컬렉션 활용(지난번 피드백 주신 내용)
- 클래스 분리
- 매직넘버 제거
- 각 객체에게 메세지를 보내자
- TDD로 기능구현

리뷰 잘 부탁드립니다!

yoonmin-kim committed:
refactor: 명확한 의미전달 및 컬렉션 활용 요소 리팩토링

변경내용:
- CustomSplitter.java
  - group()에 사용 된 매직넘버를 상수로 변경하여 의미를 전달
- Target.java
  - 일반 배열보다는 컬렉션 프레임워크를 적극 활용(리뷰어님 피드백)
- PositiveNumberValidator.java
  - RuntimeException 보다는 의미가 명확한 IllegalArgumentException 사용
- *Test.java
  - 0과 null 을 직접 입력하기 보다 @NullAndEmptySource 사용
  - assertj 에서 제공하는 assertThatIllegalArgumentException() 사용

yoonmin-kim committed:
docs(README): 기능 목록 정리

크게 3가지 기능 구분:
- 구입 금액 검증
- 당첨 번호 검증
- 로또(자동) 기능

yoonmin-kim committed:
feat(purchase): 구입 금액이 숫자인지 확인

검증 기능:
- 숫자라면 통과
- 숫자가 아니면 IllegalArgumentException 발생

yoonmin-kim committed:
feat(purchase): 구입 금액이 1000원 이상 인지 확인

검증 기능:
- 1000원 이상이라면 통과
- 1000원 미만이면 IllegalArgumentException 발생

yoonmin-kim committed:
feat(purchase): 구입 금액이 1000원 으로 나누어 떨어지는지 확인

검증 기능:
- 1000원으로 나누어 떨어지면 통과
- 1000원으로 나누어 떨어지지 않으면 IllegalArgumentException 발생

yoonmin-kim committed:
feat(winningnumber): 쉼표(,)로 split 한 값이 6개인지 확인

검증 기능:
- 값이 6개면 통과
- 값이 6개가 아니면 IllegalArgumentException 발생

yoonmin-kim committed:
feat(winningnumber): 쉼표(,)로 split 한 값이 6개인지 확인

검증 기능:
- 값이 6개면 통과
- 값이 6개가 아니면 IllegalArgumentException 발생

yoonmin-kim committed:
feat(winningnumber): 쉼표(,)로 split 한 값이 1 ~ 45 범위인지 확인

검증 기능:
- 값이 1~45 사이에 포함되면 통과
- 값이 1~45 사이에 포함되지 않으면 IllegalArgumentException 발생

yoonmin-kim committed:
refactor(winningnumber): 쉼표(,)로 split 한 값에 중복이 있는지 확인

수정 내용:
- List 를 Set 으로 변경하여 값이 6개가 되지 않을 경우 검증 실패

yoonmin-kim committed:
docs(README): 로또번호 생성 기능을 더 작게 분리

내용:
- 생성 할 로또번호의 갯수를 먼저 구한다
- 갯수 만큼 로또번호를 생성한다

yoonmin-kim committed:
refactor: purchase 패키지를 lottonumber 하위 패키지로 이동

yoonmin-kim committed:
feat(purchase): 구입 금액을 1000원 으로 나눈 수 만큼 로또번호 갯수 생성

기능:
- LottoNumberCountMaker.java 에 구현

yoonmin-kim committed:
feat(lottonumber): 로또번호 갯수 만큼 로또번호 생성

내용:
- 로또번호 갯수 만큼 로또번호를 생성 한다
- 각 로또번호는 1 ~ 45 범위의 겹치지 않는 숫자 6개를 가진다

yoonmin-kim committed:
docs(README): 로또(자동) 기능목록에 당첨번호 기능 추가

yoonmin-kim committed:
feat(winningnumber): 입력 된 당첨번호를 쉼표(,)로 split 하여 공백을 제거한다

yoonmin-kim committed:
feat(winningnumber): 로또번호와 당첨번호를 비교하여 일치한 갯수를 확인

yoonmin-kim committed:
feat(result): 총 수익률 계산

기능 내용:
- 구입금액을 이익금으로 나누어 수익률을 계산한다
- 수익률이 1이하이면 손해라는 문구를 덧붙인다

yoonmin-kim committed:
refactor: 도메인 기능을 domain 패키지 하위로 전부 이동

yoonmin-kim committed:
docs(README): 당첨 번호 검증기능 목록 추가

yoonmin-kim committed:
feat(lotto): 로또기능 최종 개발 완료
내용:
- 사용자의 입력 및 결과 출력은 console 로 진행한다
- 잘못 된 입력을 받을 경우 해당 값을 다시 입력받는다

wheejuni commented on src/main/java/calculator/domain/target/Target.java:
```diff
@@ -18,15 +21,15 @@ public Target(String target) {
+        String[] splits = splitter.split(target);
+        List<Integer> result = new ArrayList<>();
+        for (String split : splits) {
+            result.add(isValidatedThenParseInt(split));
```
스트림 문법으로 개선을 시도해주세요

yoonmin-kim replied:
스트림을 활용하여 간결하게 표현할 수 있도록 수정하였습니다

wheejuni commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,57 @@
+        while (!isComplete) {
+            isComplete = readPurchase();
+        }
+    }
```
프라이빗 메소드 추출로 인해 오히려 가독성이 떨어졌다는 느낌이 들고요...
while loop와 `readPurchase()` 가 통합되어 있으면 좋겠다는 생각이 듭니다.

yoonmin-kim replied:
가독성 있게 수정하면서 코드의 중복도 제거하기 위한 시도를 해보았습니다

wheejuni commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,31 @@
+import lotto.domain.result.DefaultLottoResult;
+import lotto.domain.winningnumber.WinningNumber;
+
+public class Lotto {
```
클래스 이름이 명확하지 않다고 생각합니다. 사용자가 진행하는 게임, 그러니까 여러장의 로또 티켓의 통합체인지? 어떤 정보들을 관리할 클래스인지 조금 더 명확하면 좋을 것 같군요.

yoonmin-kim replied:
클래스명을 역할을 떠올릴 수 있도록 `LottoNumberMatcher`로 수정했습니다

wheejuni commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,31 @@
+    public DefaultLottoResult result() {
+        for (Iterator<Integer> lottoNumberIterator : lottoNumbers.createIterators()) {
+            lottoResult.calculateTotalCount(winningNumber.matchNumber(lottoNumberIterator));
+        }
```
`Iterator` 를 순회하는 개념이 그렇게 직관적이진 않은 것 같습니다.
`LottoNumbers` 에 `WinningNumber` 와 비교해 자기 자신의 등수를 알려주는 로직이 존재하면 더 간결할 것 같고,
Stream API의 사용으로도 리팩터링 가능할 것 같은데요.

yoonmin-kim replied:
* `Iterator`를 가져 올 필요 없이 상태를 갖고 있는 `LottoNumbers`에 메세지를 보내도록 수정하였습니다
* 해당 메서드에서는 Stream을 사용하였습니다

wheejuni commented on src/main/java/lotto/domain/lottonumber/LottoNumber.java:
```diff
@@ -0,0 +1,51 @@
+    private static final int RESULT_START_INDEX = 0;
+    private static final int RESULT_END_INDEX = (-2);
+    private static final int CREATE_START_INDEX = 0;
+    private static final int CREATE_END_INDEX = 6;
```
과한 추출이라고 생각합니다. `private` 이라고 한정한 것 자체가 이 클래스에서만 사용하겠다는 것인데,
이 정도는 인라인으로 그냥 명시해도 충분하다고 생각합니다.

yoonmin-kim replied:
과하게 추출 된 상수들을 모두 제거하였습니다

wheejuni commented on src/main/java/lotto/domain/lottonumber/LottoNumber.java:
```diff
@@ -0,0 +1,51 @@
+        Collections.sort(lottoNumber);
+    }
+
+    public Iterator<Integer> createIterator() {
+        return lottoNumber.iterator();
+    }
```
`Iterator` 를 직접 사용해야 하는 경우에 대해 그림이 잘 그려지지 않습니다. 왜 이런 필요가 생겼는지 다시 한번 생각해 보시면 어떨까요?
대개 이런 경우에 이 `Iterator` 를 받아다 사용하는 로직을 여기 클래스에 바로 위치시키는 것이 낫습니다.

wheejuni commented on src/main/java/lotto/domain/lottonumber/LottoNumbers.java:
```diff
@@ -0,0 +1,36 @@
+    private static final int MAKE_START_INDEX = 0;
+    private List<LottoNumber> lottoNumbers = new ArrayList<>();
+
+    public LottoNumbers(Purchase purchase) {
+        int lottoNumberCount = purchase.makeLottoNumberCount();
+        for (int i = MAKE_START_INDEX; i < lottoNumberCount; i++) {
+            lottoNumbers.add(new LottoNumber());
+        }
+    }
```
생성자가 좀 더 간단해져야 한다고 생각합니다.
별도의 스태틱 팩토리 메서드나, `LottoNumbersGenerator` 같은 것으로 분리해야 할 것 같아요.

wheejuni commented on src/main/java/lotto/domain/lottonumber/LottoNumbers.java:
```diff
@@ -0,0 +1,36 @@
+        }
+    }
+
+    public List<Iterator<Integer>> createIterators() {
+        List<Iterator<Integer>> iterators = new ArrayList<>();
+        lottoNumbers.forEach((lottoNumber) -> iterators.add(lottoNumber.createIterator()));
+        return iterators;
+    }
```
이 메소드가 리턴하는 리스트가 어떻게 활용될지 감이 잘 잡히지 않고, 그닥 좋은 방법도 아니라고 생각합니다.
앞선 코멘트에서도 의견 드렸던 것 같은데 이렇게 `Iterator` 를 리턴 할 거라면 그 Iterator로 수행할 작업을 아예 여기에 구현하는 것도 방법입니다.

wheejuni commented on src/main/java/lotto/domain/lottonumber/purchase/Purchase.java:
```diff
@@ -0,0 +1,36 @@
+    private PurchaseRole role;
+
+    public Purchase(String purchase) {
+        new DefaultPurchaseValidator().validate(purchase);
+        this.purchase = purchase;
+        this.role = new LottoNumberCountMaker();
```
생성자는 항상 간결하게 가져가 주세요.

yoonmin-kim replied:
명심하겠습니다. 세심한 의견 감사드립니다!

wheejuni commented on src/main/java/lotto/domain/lottonumber/purchase/validation/PurchaseValidator.java:
```diff
@@ -0,0 +1,10 @@
+package lotto.domain.lottonumber.purchase.validation;
+
+public interface PurchaseValidator {
```
이 인터페이스의 추상화에 공감이 잘 가지 않습니다.
`validate` 라는 메서드가 `void` 리턴을 갖는데 이것이 `Purchase` 에 국한되는 것인지 이해하기 어렵습니다.
차라리 `String purchase` 를 받아 실제 `Purchase` 객체를 만들어내는 로직이 담겨 있었다면 조금 더 공감이 갈 것 같습니다.

yoonmin-kim replied:
* `Purchase` 를 생성해주는 별도의 `PurchaseFactory`를 추가하였습니다
* `PurchaseFactory` 내부에서 `validator` 를 활용합니다

wheejuni commented on src/main/java/lotto/domain/lottonumber/purchase/validation/PurchaseValidator.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface PurchaseValidator {
+
+    String ERROR_NUMBER_MESSAGE = "[ERROR] 구입 금액은 숫자로만 입력해주세요.";
+    String ERROR_MIN_COST_MESSAGE = "[ERROR] 구입 가능 금액은 최소 1000원 입니다";
+    String ERROR_REMAINDER_MESSAGE = "[ERROR] 구입 금액은 1000원 단위로만 입력 가능합니다.";
```
인터페이스에 상수를 보관하는 것은 [그리 좋은 패턴은 아닙니다. ](https://jinseongsoft.tistory.com/190)

yoonmin-kim replied:
인터페이스만 보고 기능을 판단할 수 있도록 해보면 어떨까? 라는 의도로 했던 작업인데 생각해보니 같은 패키지 클래스만 보고도 파악이 가능하기 때문에 큰 의미가 없다는것을 깨달았습니다. 

좋은 정보(링크)도 감사드립니다!

wheejuni commented on src/main/java/lotto/domain/repository/LottoNumbersRepositoryImpl.java:
```diff
@@ -0,0 +1,18 @@
+
+import lotto.domain.lottonumber.LottoNumbers;
+
+public class LottoNumbersRepositoryImpl implements LottoNumbersRepository {
```
지금은 콘솔 전용 앱이라 그럴 일은 없겠습니다만 - 스레드 경합 상황이 발생했을 때 이 객체는 설계대로 작동할까요?

yoonmin-kim replied:
[수정사항]
* 피드백 주신 내용을 듣고 `Repository`를 제거 후 `LottoGame`객체 내부에서 `LottoNumbers`의 상태를 갖게 했습니다

[수정사항과 별개로 생각해본 내용]
* `메모리 Repository`를 유지한다면 `ThreadLocal`을 활용하여 `LottoNumbers`를 저장해야 할 것 같습니다
* 멀티쓰레드가 고려사항이 아니다 보니 `LottoGame`을 싱글턴으로 구현하였는데요
* 만약 멀티쓰레드 환경으로 변경 되어야 한다면 `LottoGame`를 각 쓰레드 마다 생성하도록 수정해야 할 것 같습니다

wheejuni reviewed (changes requested):
테스트 케이스를 꼼꼼히 작성하시고, 새로운 시도를 가져가 보신 점은 좋은 것 같습니다. 💯 
그러나 자유롭게 패턴을 가져가며 많은 시도를 할 수 있는 콘솔 기반 앱 프로젝트에서 특정 프레임워크/패턴의 방법론이 지나치게 많이 등장한 점은 그리 좋은 평가를 드릴 수는 없을 것 같습니다.

일회성으로 실행되는 콘솔 기반 게임이고 데이터가 영속화되지 않기 때문에 `Repository` 는 필요 없다고 단언할 수 있습니다. 좋은 시도였지만, 4단계에선 조금은 다른 방향의 설계를 기대해 보겠습니다. 👍

yoonmin-kim committed:
refactor: 클래스명 변경 및 스트림 문법으로 코드를 간결하게 표현

yoonmin-kim committed:
refactor: Controller 에 가독성 떨어지는 코드수정

yoonmin-kim committed:
refactor: 과한 상수제거 및 의미있는 메소드 생성

yoonmin-kim committed:
refactor: Repository 제거 및 메소드명 명확하게 수정

yoonmin-kim committed:
refactor: WinningNumberFactory 추가 및 검증위치 변경

내용:
- Factory 를 통해서 WinningNumber 를 생성한다
- 기존에 WinningNumber 에 수행하던 검증기능을 Factory
에서 수행한다

wheejuni reviewed (approved):
이것저것 변경을 많이 요청드렸는데 빠르게 반영해 주셨군요. 수고 많으셨습니다. 👍
